### DISCUSSION TITLE
How to implement door/elevator "squish" behavior?

### DISCUSSION THREAD

**Discussion by demiurghg** [bepu/bepuphysics1#24] [Q&A] [answered]
Hi!

I have door or elevator (kinematic convex hull with attached linear and angular motor that moves it up and down).

How handle situation when player (character controller) or other entity (box, barrel) get stuck in door or elevator and prevents them from moving? 

Naive implementation No1 just pushes object through the static geometry
Naive implementation No2 with handling collision events help detect situation when player touches the door or elevator, but I have no idea how to handle situation when player steps on the elevator and when elevator goes up, squish player against the ceiling. See attached image.

![image](https://user-images.githubusercontent.com/3645657/142514998-966b1ad2-3a6d-4620-b135-68b060b8325c.png)

**Questions:**
How to detect situation when player or dynamic object prevent kinematic object from moving?
How to detect situation when player or dynamic object got out and door/elevator may continue movement?

**Comment by RossNordby** [comment, marked as answer]
One approach would be to examine the accumulated impulses associated with the contact constraints on the kinematic elevator. If contact contact constraints are applying a lot of force to the kinematic, it is probably squishing something (or the object's just very heavy). You could stop the kinematic's progress or even reverse it a little bit in response.

After a little while of not detecting the excessive impulse, you could return it to the usual movement pattern.

You could also make the elevator a dynamic body with constraints that make it act elevator-y; the motor raising and lowering the elevator could be given a maximum force that doesn't make it smash stuff through walls. Depending on your goals, this might be the easiest approach, especially for doors.

**Reply by demiurghg** [reply]
Performed several tests. Approach with tracking of the contact normal and tangential impulses looks pretty promising.

However, there is a problem: the impulse that squashes player through the floor or ceiling is about 500–1000 units (for character with mass 10). But in the same time, when player simply jumps and lands on the elevator the contact impulse have similar order of magnitude. 

Investigating, I've found `Pair.Contact.Contact.PenetrationDepth` that seems could be used as more measurable value to apply damage. I.e. when penetration depth is greater than 30% of minimal size of the object we damage object and stop the door or elevator. 

But even using such approach its hard to tell door squashes the player or player smashes himself against the door. Could you suggest way to tell both cases? (Yes, reality is cruel and from physics point of view both cases are similarly destructive, but for game I'd prefer to give more freedom :)

**Reply by RossNordby** [reply]
Some other bits of information you could consider:
1. Previous velocities of bodies. If the previous incoming velocity wasn't high but the contact has a high impulse, it is more likely to be squishing. It'd be up to you to track those velocities.
2. User input. If you know the player is doing something weird, you could make the world respond differently.
3. Contact information about the candidate object. For example, if your elevator or door detects that it's applying a large impulse to a body through a contact, then you could look at that body's contacts in turn- if the potential squishvictim has a large contact impulse with *another* object that isn't the elevator/door, then it's more likely that it's getting squished (as opposed to just running and jumping around).

**Reply by demiurghg** [reply]
Option No. 3 seems the best solution. 

Formerly, I supposed that door/elevator system should be responsible for damage.

But approach No. 3 allows to create separate system that controls contact pair, detects squish-victims, applies damage and sends event to stop doors/elevator. Such approach seems more general and allows to track squash-damage not only by door/elevators, but from any other physical objects, (i.e. player got squashed by free falling container or pile of ragdolls).

Thank you for advice!
